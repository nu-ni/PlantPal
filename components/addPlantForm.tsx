import React, { useRef, useState } from "react";
import {
  View,
  TextInput,
  Button,
  Text,
  StyleSheet,
  TouchableOpacity,
  Image,
  Alert,
} from "react-native";
import { Ionicons } from "@expo/vector-icons";
import { CameraType, CameraView } from "expo-camera";
import * as ImagePicker from "expo-image-picker";
import {
  getLastActiveCollection,
  insertData,
} from "@/services/DatabaseService";
import { Plant } from "@/data/models";
import { ActionButton } from "./actionButtton";

export function AddPlantForm({
  onBackButtonClick: onBackButtonClick,
}: {
  onBackButtonClick: () => void;
}) {
  const [plantName, setPlantName] = useState<string>("");
  const [timesPerWeek, setTimesPerWeek] = useState<number>(0);
  const [amount, setAmount] = useState<number>(0);
  const [plantImage, setPlantImage] = useState<string>("");
  const [cameraScreenVisible, setCameraScreenVisible] =
    useState<boolean>(false);
  const [facing, setFacing] = useState<CameraType>("back");
  const cameraRef = useRef<CameraView>(null);

  const handlePlantNameChange = (text: string) => {
    setPlantName(text);
  };

  const handleTimesPerWeekChange = (text: string) => {
    setTimesPerWeek(Number(text));
  };

  const handleAmountChange = (text: string) => {
    setAmount(Number(text));
  };

  const handleCameraIconClick = () => {
    setCameraScreenVisible((prevState) => !prevState);
  };

  const toggleCameraFacing = () => {
    setFacing((current) => (current === "back" ? "front" : "back"));
  };

  const takePhoto = async () => {
    if (cameraRef.current) {
      const photo = await cameraRef.current.takePictureAsync({ base64: true });
      const base64String = photo.base64 || "";
      setPlantImage(base64String);
    }
    setCameraScreenVisible(false);
  };

  const selectImageFromGallery = async () => {
    const result = await ImagePicker.launchImageLibraryAsync({
      mediaTypes: ImagePicker.MediaTypeOptions.Images,
      aspect: [4, 3],
      quality: 1,
      base64: true,
    });

    if (!result.canceled && result.assets[0].base64) {
      setPlantImage(result.assets[0].base64);
    }
  };

  const removeImage = () => {
    setPlantImage("");
  };

  const handleSubmit = async () => {
    if (plantName.trim() === "") {
      Alert.alert("Validation Error", "Plant Name cannot be empty.");
      return;
    }
    if (plantImage.trim() === "") {
      Alert.alert("Validation Error", "Please add an image of the plant.");
      return;
    }

    const lastActiveCollection = await getLastActiveCollection();
    if (lastActiveCollection) {
      let tableName = "Plant";
      let data: Plant = {
        title: plantName,
        frequency: timesPerWeek,
        waterAmount: amount,
        collectionId: lastActiveCollection[0].id,
        image: plantImage,
      };

      await insertData(tableName, data);

    }
    onBackButtonClick();
  };

  return (
    <View>
      <View style={styles.row}>
        <Text style={styles.label}>Plant Name:</Text>
        <TextInput
          style={styles.input}
          value={plantName}
          onChangeText={handlePlantNameChange}
        />
      </View>
      <View style={styles.row}>
        <Text style={styles.label}>Times Per Week:</Text>
        <TextInput
          style={styles.input}
          value={timesPerWeek.toString()}
          onChangeText={handleTimesPerWeekChange}
          keyboardType="numeric"
        />
      </View>

      <View style={styles.row}>
        <Text style={styles.label}>Amount:</Text>
        <TextInput
          style={styles.input}
          value={amount.toString()}
          onChangeText={handleAmountChange}
          keyboardType="numeric"
        />
      </View>

      <View style={styles.row}>
        <Text style={styles.label}>Plant Image:</Text>
      </View>

      {!cameraScreenVisible && (
        <View style={styles.iconContainer}>
          <Ionicons name="camera" size={30} onPress={handleCameraIconClick} />
          <TouchableOpacity onPress={selectImageFromGallery}>
            <Ionicons name="image" size={30} />
          </TouchableOpacity>
        </View>
      )}

      {cameraScreenVisible && (
        <View>
          <Ionicons name="close" size={30} onPress={handleCameraIconClick} />
          <View style={styles.container}>
            <CameraView ref={cameraRef} style={styles.camera} facing={facing}>
              <View style={styles.buttonContainer}>
                <TouchableOpacity
                  style={styles.button}
                  onPress={toggleCameraFacing}
                >
                  <Text style={styles.text}>Flip Camera</Text>
                </TouchableOpacity>
                <TouchableOpacity style={styles.button} onPress={takePhoto}>
                  <Text style={styles.text}>Take Photo</Text>
                </TouchableOpacity>
              </View>
            </CameraView>
          </View>
        </View>
      )}

      {plantImage ? (
        <View>
          <Image
            source={{ uri: `data:image/jpeg;base64,${plantImage}` }}
            style={{ width: 100, height: 100 }}
          />
          <Ionicons
            name="close-circle"
            size={30}
            onPress={removeImage}
            style={{ position: "absolute", top: 0, left: 60 }}
          />
        </View>
      ) : null}
      <ActionButton style={styles.actionButton} title="Save" onPress={handleSubmit} />
      <ActionButton style={styles.actionButton} title="Back" onPress={onBackButtonClick} />
    </View>
  );
}

const styles = StyleSheet.create({
  actionButton: {
  marginTop: 20
  },
  row: {
    flexDirection: "row",
    alignItems: "center",
    marginBottom: 10,
  },
  label: {
    color: "#E3E3E3",
    marginRight: 10,
    fontSize: 16,
    fontWeight: "bold",
  },
  input: {
    flex: 1,
    backgroundColor: "white",
    height: 40,
    borderColor: "gray",
    borderWidth: 1,
    borderRadius: 5,
    paddingHorizontal: 10,
  },
  container: {
    flex: 1,
    justifyContent: "center",
  },
  camera: {
    flex: 1,
  },
  buttonContainer: {
    flex: 1,
    flexDirection: "row",
    backgroundColor: "transparent",
    margin: 64,
  },
  button: {
    flex: 1,
    alignSelf: "flex-end",
    alignItems: "center",
  },
  text: {
    fontSize: 24,
    fontWeight: "bold",
    color: "white",
  },
  iconContainer: {
    flexDirection: "row",
    justifyContent: "space-around",
    marginVertical: 10,
  },
});
