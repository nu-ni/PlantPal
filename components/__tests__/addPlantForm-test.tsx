import React from "react";
import { AddPlantForm } from "../addPlantForm";
import renderer from "react-test-renderer";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import { Alert } from "react-native";

// TODO: snapshot
it("renders correctly", () => {
  const tree = renderer
    .create(
      <AddPlantForm
        onButtonClick={function (): void {
          throw new Error("Function not implemented.");
        }}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

// postitive
it("should show camera view when the camera icon is clicked", () => {
  const { getByText, getByTestId } = render(
    <AddPlantForm onButtonClick={jest.fn()} />
  );

  const cameraIcon = getByTestId("camera-icon");
  fireEvent.press(cameraIcon);

  expect(getByText("Flip Camera")).toBeTruthy();
  expect(getByText("Take Photo")).toBeTruthy();
});

// negative
it("should display an alert when plant name is empty", async () => {
  jest.spyOn(Alert, "alert");
  const { getByText } = render(<AddPlantForm onButtonClick={jest.fn()} />);

  const saveButton = getByText("Save");
  fireEvent.press(saveButton);

  await waitFor(() => {
    expect(Alert.alert).toHaveBeenCalledWith(
      "Validation Error",
      "Plant Name cannot be empty."
    );
  });
});