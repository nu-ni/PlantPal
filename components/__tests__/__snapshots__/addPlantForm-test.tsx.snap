// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 10,
      }
    }
  >
    <Text
      style={
        {
          "color": "#E3E3E3",
          "fontSize": 16,
          "fontWeight": "bold",
          "marginRight": 10,
        }
      }
    >
      Plant Name:
    </Text>
    <TextInput
      onChangeText={[Function]}
      style={
        {
          "backgroundColor": "white",
          "borderColor": "gray",
          "borderRadius": 5,
          "borderWidth": 1,
          "flex": 1,
          "height": 40,
          "paddingHorizontal": 10,
        }
      }
      value=""
    />
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 10,
      }
    }
  >
    <Text
      style={
        {
          "color": "#E3E3E3",
          "fontSize": 16,
          "fontWeight": "bold",
          "marginRight": 10,
        }
      }
    >
      Times Per Week:
    </Text>
    <TextInput
      keyboardType="numeric"
      onChangeText={[Function]}
      style={
        {
          "backgroundColor": "white",
          "borderColor": "gray",
          "borderRadius": 5,
          "borderWidth": 1,
          "flex": 1,
          "height": 40,
          "paddingHorizontal": 10,
        }
      }
      value="0"
    />
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 10,
      }
    }
  >
    <Text
      style={
        {
          "color": "#E3E3E3",
          "fontSize": 16,
          "fontWeight": "bold",
          "marginRight": 10,
        }
      }
    >
      Amount:
    </Text>
    <TextInput
      keyboardType="numeric"
      onChangeText={[Function]}
      style={
        {
          "backgroundColor": "white",
          "borderColor": "gray",
          "borderRadius": 5,
          "borderWidth": 1,
          "flex": 1,
          "height": 40,
          "paddingHorizontal": 10,
        }
      }
      value="0"
    />
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 10,
      }
    }
  >
    <Text
      style={
        {
          "color": "#E3E3E3",
          "fontSize": 16,
          "fontWeight": "bold",
          "marginRight": 10,
        }
      }
    >
      Plant Image:
    </Text>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-around",
        "marginVertical": 10,
      }
    }
  >
    <Text />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text />
    </View>
  </View>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderRadius": 10,
            "padding": 10,
          },
          {
            "marginTop": 20,
          },
          false,
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "black",
            },
            false,
          ]
        }
      >
        Save
      </Text>
    </View>
  </View>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderRadius": 10,
            "padding": 10,
          },
          {
            "marginTop": 20,
          },
          false,
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "black",
            },
            false,
          ]
        }
      >
        Back
      </Text>
    </View>
  </View>
</View>
`;
